### ISSUE TITLE
Divergent code paths for forces and single particle energies

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#1870]
Work on #1869 revealed that the `evaluate` functionality in `QMCHamiltonianBase` objects is highly fragmented when selecting among the following four options:

1. No forces, no single particle energies (the default).
2. Forces, no single particle energies.
3. No forces, single particle energies.
4. Forces and single particle energies (not yet supported).

Currently a separate evaluate function exists for each of 1-3.  A fourth would need to be added to support 4, and we don't want to get to the case where another quantity similar in need to forces or single particle energies is added (9 evaluate functions).

I propose we merge the functionality 1-4 into a single evaluate function with appropriate dynamic logic.  The fact that some logic is already present in these functions obviates concerns about any significant performance hit from the merger.

**Comment by ye-luo** [comment]
I feel the requirement for force evaluation is different from energy only calculation. Force is a estimator depends on the Hamiltonian derivative operation and it doesn't need to be evaluated as frequent as energy.
So I prefer a separation. H.evaluate() does 1 and 3. H.evaluateIonDeriv() for the force only.

**Comment by jtkrogel** [comment]
Both the force and the energy density are semi-local observables that do not commute with the Hamiltonian, but F, E(r) and H all share underlying components.  For observation purposes, they likely all need to be evaluated just as frequently as each other.

Quite apart from this is the code divergence issue.  I don't disagree about having H.evaluateIonDeriv(), but inside it should call only H.evaluate(ion_deriv=true), i.e. there is a single implementation that is merged to the degree possible (lower divergence) with a little cheap logic as the price.

**Comment by ye-luo** [comment]
@jtkrogel I'm OK with evaluate_impl(a few switches). Then H.evaluate/evaluateIonDerivs  goes through it.

**Comment by rcclay** [comment]
To what extent is code encapsulation worth the amount of duplication/divergence?  The main issue I see is that when you want to add a new derivative or observable, you add a new "evaluate" function, as opposed to adding in if(newFeature){} block in the main energy "evaluate" block.  That last one seems like it could get unweildy.

**Comment by jtkrogel** [comment]
It's in the combination of functionality that you get killed.  For example forces + single particle energies requires the addition of a 4th function.  With forces + sp energies + something new you get 9 functions, then 16, 25, etc.  The divergence makes the code size balloon with duplication and makes maintenance much more difficult.

**Comment by prckent** [comment]
A design that can not contain or avoid the combinatorial explosion outlined by Jaron is a failed design.

**Comment by markdewing** [comment]
Just to be clear, what are the evaluation functions for 1-3?  My guess is
1. evaluate
2. evaluateWithIonDerivs
3. evaluate_sp

**Comment by jtkrogel** [comment]
Yes, those are the current functions.  The 4th would be something like evaluateWithIonDerivs_sp (violations to naming convention aside).

**Comment by PDoakORNL** [comment]
I think its worth exploring separating out the state and results from the computation. Then we could define a set of evalutation types clear about their input arguments. 

Using these evaluation types we could define a sort order usable for a vector of variants (a graph would be optimal). The evaluation visit to each "element" of the Hamiltonian could use previous computed results. Adding a new evaluation would require adding an eval call to it for the visitor and supplying it with previous computed input if necessary (maybe a little like the if discussed above). Additionally if computation dependency on computation not previously in the Hamiltonian element vector was needed an "default" element that produced it would be needed.

I think this might offer much better encapsulation plus the expression of dependency which we try to express through the inheritance hierarchy now but which has become quite unwieldly.

**Comment by rcclay** [comment]
@jtkrogel @prckent But it's not really combinatorial if you don't mind a little bit of redundant calculation...  for example by changing evaluateWithIonDerivs_sp -> evaluate_sp; evaluateWithIonDerivs;.  Energy+force calculations are certainly lumped together in evaluateWithIonDerivs because it's convenient to have the intermediates that come about during the energy calculations when computing the forces...  but the cost of recomputing these intermediates is much less than the total force computation.  If it looks like the performance savings of computing these intermediates only once over all observables is worth it, it might be worth seriously reconsidering how the class and API's are designed.  For example, making the intermediates class members, or having the evaluates take the intermediates as an argument would be a first step to separating state and results of computation as Peter says.  Both evaluateXandYandZ() methods and one monster evaluate() with blocks like if(x)...else if(y)...else if (z) seem very nonideal, and have been used quite liberally in QMCPACK
